src/watch_filter.rs

```diff
@@ -173,11 +200,68 @@ impl Filterer for WatchFilter {
         }
 
         // By process of elimination the event is valid
-        tracing::trace!("passing event: {event:?}");
+        tracing::debug!("passing event: {event:?}");
         Ok(true)
     }
 }
 
+#[derive(Debug)]
+pub struct Config {
+    pub manifest_pattern: WatchPattern,
+    pub source_patterns: Vec<WatchPattern>,
+    pub artifact_patterns: Vec<WatchPattern>,
+    pub ignore_patterns: Vec<WatchPattern>,
+}
+
+pub struct WatchPattern {
+    pub active: bool,
+    pub glob: String,
+    pub pattern: Pattern,
+}
+
+impl fmt::Debug for WatchPattern {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "WatchPattern {{ glob: {:?}, active: {:?} }}",
+            self.glob, self.active
+        )
+    }
+}
+
+#[derive(Debug)]
+pub struct WatchPatternBuilder<'a> {
+    app_dir: &'a PathBuf,
+}
+
+impl<'a> WatchPatternBuilder<'a> {
+    pub fn new(app_dir: &'a PathBuf) -> Self {
+        tracing::debug!("building watch patterns relative to app dir: {app_dir:?}");
+        Self { app_dir }
+    }
+
+    pub fn watched(&self, glob: String) -> Result<WatchPattern> {
+        self.build(glob, true)
+    }
+
+    pub fn not_watched(&self, glob: String) -> Result<WatchPattern> {
+        self.build(glob, false)
+    }
```

Comment:
I get the motivation for doing this, but given that these are only called inside boolean conditional expressions I'm pretty skeptical that this is a net positive. A single e.g. `WatchPattern::new(app_dir, glob, active)` method would cut quite a bit of code and probably be easier to read overall.